pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,440 @@
+}
+
+func NewSAToSCCCache(rbacInformers rbacv1informers.Interface, sccInfomer securityv1informers.SecurityContextConstraintsInformer) *SAToSCCCache {
+	return &SAToSCCCache{
+		roleLister:                rbacInformers.Roles().Lister(),
+		clusterRoleLister:         rbacInformers.ClusterRoles().Lister(),
+		roleBindingIndexer:        rbacInformers.RoleBindings().Informer().GetIndexer(),
+		clusterRoleBindingIndexer: rbacInformers.ClusterRoleBindings().Informer().GetIndexer(),
+
+		sccLister: sccInfomer.Lister(),
+
+		usefulRoles: make(map[string][]string),
+	}
+}
+
+// SCCsFor returns a slice of all the SCCs that a given service account can use
+// to run pods in its namespace.
+// It expects the serviceAccount name in the system:serviceaccount:<ns>:<name> form.
+func (c *SAToSCCCache) SCCsFor(serviceAccount *corev1.ServiceAccount) (sets.String, error) {
+	saUserInfo := serviceaccount.UserInfo(
+		serviceAccount.Namespace,
+		serviceAccount.Name,
+		string(serviceAccount.UID),
+	)
+
+	// realSAUserInfo adds the "system:authenticated" group to SA UserInfo groups
+	realSAUserInfo := &user.DefaultInfo{
+		Name:   saUserInfo.GetName(),
+		Groups: append(saUserInfo.GetGroups(), user.AllAuthenticated),
+		UID:    saUserInfo.GetUID(),
+		Extra:  saUserInfo.GetExtra(),
+	}
+
+	bindingObjs, err := getIndexedRolebindings(c.roleBindingIndexer, realSAUserInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterBindingObjs, err := getIndexedRolebindings(c.clusterRoleBindingIndexer, realSAUserInfo)
+	if err != nil {
+		return nil, err
+	}
+	objs := append(bindingObjs, clusterBindingObjs...)
+
+	sccs, err := c.sccLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	allowedSCCs := sets.NewString()
+	for _, scc := range sccs {
+		if sccAllowsSA(scc, realSAUserInfo) {
+			allowedSCCs.Insert(scc.Name)
+		}
+	}
+
+	for _, o := range objs {
+		rb, err := newRoleBindingObj(o)
+		if err != nil {
+			// this would be rather weird
+			return nil, err
+		}
+
+		roleCachedKey := fmt.Sprintf("/%s", rb.RoleRef().Name)
+		if rb.RoleRef().Kind == "Role" {
+			roleCachedKey = rb.Namespace() + roleCachedKey
+		}
+
+		c.usefulRolesLock.Lock()
+		// this role does not have SCC-related rules
+		cachedAllowedSCCs := c.usefulRoles[roleCachedKey]
+		c.usefulRolesLock.Unlock()
+		if len(cachedAllowedSCCs) == 0 {
+			continue
+		}
+
+		// we particularly care only about Roles in the SA NS
+		if roleRef := rb.RoleRef(); rb.AppliesToNS(serviceAccount.Namespace) && roleRef.APIGroup == rbacv1.GroupName {
+			allowedSCCs.Insert(cachedAllowedSCCs...)
+		}
+	}
+
+	return allowedSCCs, nil
+}
+
+// getRoleFromRoleRef tries to retrieve the role or clusterrole from roleRef.
+func (c *SAToSCCCache) getRoleFromRoleRef(ns string, roleRef rbacv1.RoleRef) (*roleObj, error) {
+	var role interface{}
+	var retryErr error
+	switch kind := roleRef.Kind; kind {
+	case "Role":
+		retryErr = retry.OnError(retry.DefaultBackoff, apierrors.IsNotFound, func() error {
```

Comment:
This looks questionable.  I would instead rely on being retriggered.  otherwise a set of bad rolebindings can really mess you up.

Reply:
:+1: